Consolidate accounts onboarding: Create new Google Ads account via the combo card

ankitrox opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #2603 

_Replace this with a good description of your changes & reasoning._


### Screenshots:

<!--- Optional --->


## Detailed test instructions.

Note: As this PR also depends on claim card which is being handled in #2582, an [integration branch](https://github.com/woocommerce/google-listings-and-ads/tree/2603-test-integration) is created to test this feature.

1. Checkout to the integration branch and build. Make sure you are disconnected from all Google accounts.

2. Connect to the Google account with existing Google ads accounts so that the existing ads account is shown.

3. Click on "Or, create a new Google Ads account" link. It should show you the modal.

4. Ensure that clicking on "Cancel" button does nothing.

5. Clicking on "Yes, I want a new account" should create a new account. While creating the new account, you shall see the text "Creating new ads account" in place of existing ads account card.

6. Once the account is created successfully, existing ads account card will disappear and claim card will be shown.

7. Once the ads account is claimed, you shall see the notice `Google Ads conversion measurement has been set up for your store.` in green color.

8. Existing ads account card will again be shown with the `Connected` label and the newly created ads account ID in dropdown.


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

ankitrox committed:
Create new ads account.

ankitrox committed:
Fix: lint css.

ankitrox committed:
Merge branch 'update/2596-connect-ads-account' into update/2603-create-ads-account

ankitrox committed:
Add css.

ankitrox committed:
Do not show card if account is in unclaimed status.

ankitrox committed:
Merge branch 'update/2596-connect-ads-account' into update/2603-create-ads-account

ankitrox committed:
Add E2E tests.

ankitrox committed:
Fix: E2E issues.

ankitrox committed:
Resolve conflicts in E2E.

asvinb commented on js/src/components/google-combo-account-card/connect-ads/confirm-create-modal.js:
```diff
@@ -0,0 +1,72 @@
+import WarningIcon from '.~/components/warning-icon';
+import './confirm-create-modal.scss';
+
+const ConfirmCreateModal = ( {
```
@ankitrox Missing JSDocs for the props.

asvinb commented on js/src/components/google-combo-account-card/connect-ads/confirm-create-modal.js:
```diff
@@ -0,0 +1,72 @@
+import './confirm-create-modal.scss';
+
+const ConfirmCreateModal = ( {
+	onContinue = () => {},
```
@ankitrox No need to set a default value for the props since they will always be defined.

asvinb commented on js/src/components/google-combo-account-card/connect-ads/confirm-create-modal.js:
```diff
@@ -0,0 +1,72 @@
+					key="confirm"
+					isSecondary
+					eventName="gla_ads_account_warning_modal_confirm_button_click"
+					onClick={ handleCreateAccountClick }
```
@ankitrox I think we can call `onContinue` directly.

asvinb commented on js/src/components/google-combo-account-card/connect-ads/connect-ads-footer.js:
```diff
@@ -14,14 +15,14 @@ import DisconnectAccount from '.~/components/google-ads-account-card/disconnect-
  * @return {JSX.Element} Footer component.
  */
-const ConnectAdsFooter = ( { isConnected } ) => {
+const ConnectAdsFooter = ( { isConnected, onCreateNew = noop } ) => {
```
```suggestion
const ConnectAdsFooter = ( { isConnected, onCreateNew } ) => {
```
From the code, looks like `onCreateNew` will always be defined.

asvinb commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -8,14 +8,18 @@ import { __ } from '@wordpress/i18n';
-import ConnectAccountCard from '../connect-account-card';
-import ConnectAdsFooter from './connect-ads-footer';
-import ConnectAdsBody from './connect-ads-body';
+import LoadingLabel from '.~/components/loading-label/loading-label';
```
```suggestion
import LoadingLabel from '.~/components/loading-label';
```

asvinb commented on js/src/components/google-combo-account-card/connect-ads/useCreateAccountActions.js:
```diff
@@ -0,0 +1,41 @@
+
+const useCreateAccountActions = () => {
+	const [ creatingNewAccount, setCreatingNewAccount ] = useState( false );
+	const [ showCreateNewModal, setShowCreateNewModal ] = useState( false );
```
@ankitrox I would move the `showCreateNewModal` logic outside of the hook since the hook should only be responsible to create a new account. Having the logic to show/hide the modal should not be part of it.

ankitrox replied:
I think this hook is not at all required and we can move everything in this hook inside `connectAds` component.

asvinb commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -80,6 +94,21 @@ const ConnectAds = () => {
 		}
 	};
 
+	if ( creatingNewAccount ) {
```
@ankitrox Shall we add a condition when we don't have a Google Ads ID?

ankitrox replied:
This won't be necessary as there won't be a scenario where Google Ads id doesn't exist, but account is getting created. Automatic account creation will do that, but at that time, this card won't render.

asvinb commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -66,7 +80,7 @@ const ConnectAds = () => {
 			await fetchGoogleAdsAccountStatus();
-			await refetchGoogleAdsAccount();
+			refetchGoogleAdsAccount();
 			setLoading( false );
```
@ankitrox Can we move the `setLoading` after the try/catch statement?

ankitrox replied:
Added `setLoading( false );` inside `finally` block.

ankitrox committed:
Resolve conflicts.

ankitrox committed:
CR feedback.

ankitrox committed:
Resolve conflicts.

ankitrox committed:
Merge branch 'update/2596-connect-ads-account' into update/2603-create-ads-account

asvinb committed:
Show correct text when creating an account.

asvinb committed:
Revert change.

asvinb commented:
@joemcgill Can you kindly review the changes I made to the PR in the following commit please:
https://github.com/woocommerce/google-listings-and-ads/pull/2651/commits/3e2dcbe63d88c881ee526a773cd455a9a050733a

Thanks!

asvinb committed:
Prevent complete re rendering of the card.

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Fix error.

asvinb committed:
Fetch existing accounts again.

asvinb committed:
Display updating text only when clicking the claim button.

asvinb committed:
Save WIP.

asvinb committed:
Merge branch 'feature/2509-consolidate-google-account-cards' into update/2603-create-ads-account

asvinb committed:
Save WIP.

asvinb committed:
UX improvements.

asvinb committed:
Review comments.

asvinb committed:
Use single quotes.

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Use correct class name.

joemcgill committed:
Improve 'Connect Google account' E2E tests

This eliminates some unnecessary page.goto() calls to speed up the tests and reduce some inconsistencies.

joemcgill commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -102,34 +115,67 @@ const ConnectAds = () => {
+			<AccountCard
+				className="gla-google-combo-service-account-card--ads"
+				title={ __(
+					'Creating a new Google Ads account',
```
When `finalizeAdsAccountCreation` is `true` it would be better to show a more precise message, like "Connecting your Google Ads account".

asvinb replied:
Sure. Updated @joemcgill

joemcgill commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -102,34 +115,67 @@ const ConnectAds = () => {
+				) }
+				indicator={
+					<LoadingLabel
+						text={ __( 'Creating…', 'google-listings-and-ads' ) }
```
Similarly, if we're finalizing the account, this should probably say 'Connecting...'

asvinb replied:
👍

joemcgill commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -58,14 +58,15 @@ const ConnectedGoogleComboAccountCard = () => {
 
 	// @TODO: edit mode implementation in 2605
-	const editMode = false;
+	const editMode = true;
```
This causes the ConnectAds component to remain shown after the new account is connected, which is probably helpful for testing in development, but is confusing for QA purposes. Can this be changed back to `false`?

asvinb replied:
Sure. Updated!

joemcgill reviewed (commented):
I'm having trouble testing this accurately since I've also started hitting the errors with account creation limits but overall this is looking pretty close. I've left a few comments specific to the account creation piece.

asvinb committed:
Update title and indicator label.

asvinb committed:
Merge branch 'update/2603-create-ads-account' of github.com:woocommerce/google-listings-and-ads into update/2603-create-ads-account

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

joemcgill commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -7,44 +7,58 @@ import { __ } from '@wordpress/i18n';
  * @return {JSX.Element} {@link AccountCard} filled with content.
  */
-const ConnectAds = () => {
+const ConnectAds = ( { finalizeAdsAccountCreation } ) => {
```
I think it would be better to pass a state called `isConnecting` (or similar) here. Also, we can add two new props: `onCreate` and `isCreating` and handle the `upsertAdsAccount` logic in the parent component. The `useUpsertAdsAccount()` hook returns an `action` prop that can be used to distinguish if we are creating a new account, or updating (i.e. connecting) a new account.

Here's a nearly complete idea of how that could work:

```diff
diff --git a/js/src/components/google-combo-account-card/connect-ads/connect-ads.js b/js/src/components/google-combo-account-card/connect-ads/connect-ads.js
index f0ec7731a..20821e5e3 100644
--- a/js/src/components/google-combo-account-card/connect-ads/connect-ads.js
+++ b/js/src/components/google-combo-account-card/connect-ads/connect-ads.js
@@ -12,7 +12,6 @@ import ConnectAdsFooter from './connect-ads-footer';
 import ConfirmCreateModal from './confirm-create-modal';
 import LoadingLabel from '.~/components/loading-label';
 import useApiFetchCallback from '.~/hooks/useApiFetchCallback';
-import useUpsertAdsAccount from '.~/hooks/useUpsertAdsAccount';
 import useDispatchCoreNotices from '.~/hooks/useDispatchCoreNotices';
 import useExistingGoogleAdsAccounts from '.~/hooks/useExistingGoogleAdsAccounts';
 import useGoogleAdsAccount from '.~/hooks/useGoogleAdsAccount';
@@ -26,10 +25,12 @@ import ConnectButton from '.~/components/google-ads-account-card/connect-ads/con
  * ConnectAds component renders an account card to connect to an existing Google Ads account.
  *
  * @param {Object} props Component props.
- * @param {boolean} props.finalizeAdsAccountCreation Whether the user is in the process of finalizing the Ads account creation, i.e after the user has claimed the account and the step is conversion_action.
+ * @param {boolean} props.isConnecting Whether the user is in the process of finalizing the Ads account creation, i.e after the user has claimed the account and the step is conversion_action.
+ * @param {boolean} props.isCreating Whether the user is in the process of creating a new Ads account.
+ * @param {Function} props.onCreate A callback to fire when creating a new account.
  * @return {JSX.Element} {@link AccountCard} filled with content.
  */
-const ConnectAds = ( { finalizeAdsAccountCreation } ) => {
+const ConnectAds = ( { isConnecting, isCreating, onCreate } ) => {
 	const [ value, setValue ] = useState();
 	const [ isLoading, setLoading ] = useState( false );
 	const { createNotice } = useDispatchCoreNotices();
@@ -44,8 +45,6 @@ const ConnectAds = ( { finalizeAdsAccountCreation } ) => {
 		method: 'POST',
 		data: { id: value },
 	} );
-	const [ upsertAdsAccount, { loading: creatingNewAccount } ] =
-		useUpsertAdsAccount();
 
 	const onCreateNew = () => {
 		setShowCreateNewModal( true );
@@ -57,7 +56,7 @@ const ConnectAds = ( { finalizeAdsAccountCreation } ) => {
 
 	const handleOnContinue = async () => {
 		setShowCreateNewModal( false );
-		await upsertAdsAccount();
+		await onCreate();
 	};
 
 	useEffect( () => {
@@ -117,14 +116,14 @@ const ConnectAds = ( { finalizeAdsAccountCreation } ) => {
 
 	// Show a loading state if the Ads account is being updated or if a new Ads account is being created.
 	// If finalizeAdsAccountCreation is true, the processing is done in `ConnectedGoogleComboAccountCard`.
-	if ( creatingNewAccount || finalizeAdsAccountCreation ) {
+	if ( isConnecting || isCreating ) {
 		let title = __(
 			'Creating a new Google Ads account',
 			'google-listings-and-ads'
 		);
 		let indicatorLabel = __( 'Creating…', 'google-listings-and-ads' );
 
-		if ( finalizeAdsAccountCreation ) {
+		if ( isConnecting ) {
 			title = __(
 				'Connecting your Google Ads account',
 				'google-listings-and-ads'
diff --git a/js/src/components/google-combo-account-card/connected-google-combo-account-card.js b/js/src/components/google-combo-account-card/connected-google-combo-account-card.js
index c2866ba68..d6ba96535 100644
--- a/js/src/components/google-combo-account-card/connected-google-combo-account-card.js
+++ b/js/src/components/google-combo-account-card/connected-google-combo-account-card.js
@@ -40,7 +40,7 @@ const ConnectedGoogleComboAccountCard = () => {
 	const { invalidateResolution } = useAppDispatch();
 	const { googleAdsAccount } = useGoogleAdsAccount();
 	const { hasAccess, step } = useGoogleAdsAccountStatus();
-	const [ upsertAdsAccount ] = useUpsertAdsAccount();
+	const [ upsertAdsAccount, { loading, action } ] = useUpsertAdsAccount();
 
 	const finalizeAdsAccountCreation =
 		hasAccess === true && step === 'conversion_action';
@@ -116,7 +116,9 @@ const ConnectedGoogleComboAccountCard = () => {
 
 			{ showConnectAds && (
 				<ConnectAds
-					finalizeAdsAccountCreation={ finalizeAdsAccountCreation }
+					isConnecting={ loading && action === 'update' }
+					isCreating={ loading && action === 'create' }
+					onCreate={ () => upsertAdsAccount() }
 				/>
 			) }
 		</div>
```

asvinb replied:
That makes sense @joemcgill . Updated.

joemcgill reviewed (changes requested):
The updated indicator states look better, but I think this could be refined a bit more to remove the need to handle the account creation directly in this component and consolidate that logic in the parent component.

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Address CR feedback.

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

joemcgill reviewed (approved):
The updates look good and E2E tests are passing.

joemcgill committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

eason9487 commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -7,37 +7,62 @@ import { __ } from '@wordpress/i18n';
  * @return {JSX.Element} {@link AccountCard} filled with content.
  */
-const ConnectAds = () => {
+const ConnectAds = ( { isConnecting, isCreating, onCreate } ) => {
```
Suggestions:
- Maybe just directly pass the `action`
- Rename `onCreate` to `onRequestCreate`

Furthermore, now that this component has taken on too much functionality. Suggest splitting it up further. For example:

```jsx
const ConnectExistingAccount = ( { onCreateClick } ) => {
	const [ value, setValue ] = useState();
	const [ isLoading, setLoading ] = useState( false );
	const { createNotice } = useDispatchCoreNotices();
	const { fetchGoogleAdsAccountStatus } = useAppDispatch();
	const isConnected = useGoogleAdsAccountReady();
	const { googleAdsAccount, hasFinishedResolution, refetchGoogleAdsAccount } =
		useGoogleAdsAccount();
	const [ connectGoogleAdsAccount ] = useApiFetchCallback( {
		path: '/wc/gla/ads/accounts',
		method: 'POST',
		data: { id: value },
	} );

	useEffect( () => {
		if ( isConnected ) {
			setValue( googleAdsAccount.id );
		}
	}, [ googleAdsAccount, isConnected ] );

	const handleConnectClick = async () => {
		if ( ! value ) {
			return;
		}

		setLoading( true );
		try {
			await connectGoogleAdsAccount();
			await fetchGoogleAdsAccountStatus();
			await refetchGoogleAdsAccount();
		} catch ( error ) {
			createNotice(
				'error',
				__(
					'Unable to connect your Google Ads account. Please try again later.',
					'google-listings-and-ads'
				)
			);
		} finally {
			setLoading( false );
		}
	};

	const getIndicator = () => {
		if ( ! hasFinishedResolution ) {
			return <LoadingLabel />;
		}

		if ( isLoading ) {
			return (
				<LoadingLabel
					text={ __( 'Connecting…', 'google-listings-and-ads' ) }
				/>
			);
		}

		if ( isConnected ) {
			return <ConnectedIconLabel />;
		}

		return (
			<ConnectButton accountID={ value } onClick={ handleConnectClick } />
		);
	};

	return (
		<AccountCard
			className="gla-google-combo-account-card gla-google-combo-service-account-card--ads"
			title={ __(
				'Connect to existing Google Ads account',
				'google-listings-and-ads'
			) }
			helper={ __(
				'Required to set up conversion measurement for your store.',
				'google-listings-and-ads'
			) }
			alignIndicator="toDetail"
			indicator={ getIndicator() }
			detail={
				<AdsAccountSelectControl
					value={ value }
					onChange={ setValue }
					autoSelectFirstOption
					nonInteractive={ isConnected }
				/>
			}
			actions={
				<ConnectAdsFooter
					disabled={ isLoading }
					isConnected={ isConnected }
					onCreateNew={ onCreateClick }
				/>
			}
		/>
	);
};

const UpsertingAccount = ( { upsertingAction } ) => {
	const isConnecting = upsertingAction === 'update';

	let title = __(
		'Creating a new Google Ads account',
		'google-listings-and-ads'
	);
	let indicatorLabel = __( 'Creating…', 'google-listings-and-ads' );

	if ( isConnecting ) {
		title = __(
			'Connecting your Google Ads account',
			'google-listings-and-ads'
		);
		indicatorLabel = __( 'Connecting…', 'google-listings-and-ads' );
	}

	return (
		<AccountCard
			className="gla-google-combo-service-account-card--ads"
			title={ title }
			helper={ __(
				'This may take a few minutes, please wait a moment…',
				'google-listings-and-ads'
			) }
			indicator={ <LoadingLabel text={ indicatorLabel } /> }
		/>
	);
};

const ConnectAds = ( { onRequestCreate, upsertingAction } ) => {
	const [ showCreateNewModal, setShowCreateNewModal ] = useState( false );

	if ( upsertingAction ) {
		return <UpsertingAccount upsertingAction={ upsertingAction } />;
	}

	const handleCreateClick = () => {
		setShowCreateNewModal( true );
	};

	const handleRequestClose = () => {
		setShowCreateNewModal( false );
	};

	const handleContinue = () => {
		onRequestCreate();
		handleRequestClose();
	};

	return (
		<>
			<ConnectExistingAccount onCreateClick={ handleCreateClick } />
			{ showCreateNewModal && (
				<ConfirmCreateModal
					onContinue={ handleContinue }
					onRequestClose={ handleRequestClose }
				/>
			) }
		</>
	);
};

export default ConnectAds;
```

In this way

- It would make the code clearer and easier to handle the different statuses
- External incoming prop states could be more clearly separated from internal states, making it easier to understand the logical branches. Otherwise there are four states staggered in a single component: `isConnecting`, `isCreating`, `isLoading`, and `isConnected`.

asvinb replied:
Thanks for that. Added the components!

eason9487 commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -7,37 +7,62 @@ import { __ } from '@wordpress/i18n';
 		data: { id: value },
 	} );
 
+	const onCreateNew = () => {
```
The name of a handler would be better to start with `handle`.

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -53,22 +78,26 @@ const ConnectAds = () => {
 			await fetchGoogleAdsAccountStatus();
-			await refetchGoogleAdsAccount();
-			setLoading( false );
+			refetchGoogleAdsAccount();
```
Wonder why the `await` was removed?

asvinb replied:
This has been restored.

eason9487 commented on js/src/components/google-combo-account-card/connect-ads/connect-ads-footer.js:
```diff
@@ -14,17 +14,17 @@ import DisconnectAccount from '.~/components/google-ads-account-card/disconnect-
  *
  * @param {Object} props Props.
  * @param {boolean} props.isConnected Whether the account is connected.
+ * @param {Function} props.onCreateNew Callback to create a new account.
```
Suggest also including "click" in the name, e.g. `onCreateClick` or `onCreateNewClick`, and maybe adjusting the description to "Callback when clicking on the button to create a new account."

asvinb replied:
👍

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -591,24 +591,97 @@ test.describe( 'Set up accounts', () => {
-			test( 'should display the correct Google Ads ID when connected', async () => {
-				const googleAccountCard =
-					setUpAccountsPage.getGoogleAccountCard();
+			test( 'should send an API request to connect existing Google Ads account', async () => {
```
This teas case doesn't assert if there is an API request is sent or even assert anything else. It would be better to add the corresponding assertion.

asvinb replied:
Added assertion.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -591,24 +591,97 @@ test.describe( 'Set up accounts', () => {
-				await once.mockAdsAccountConnected();
+
 				await once.mockAdsStatusClaimed();
+				await once.fulfillAdsAccounts( {
+					id: 111111,
+				} );
```
The payload should be `Array<Object>` type.

asvinb replied:
👍

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -591,24 +591,97 @@ test.describe( 'Set up accounts', () => {
+				await expect( cancelButton ).toHaveText( 'Cancel' );
+
+				// Click the cancel button to close the modal.
+				cancelButton.click();
```
`.click()` needs `await`.

asvinb replied:
👍

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

asvinb committed:
Address CR feedback.

joemcgill committed:
Fix ConnectAds props

joemcgill committed:
Fix E2E tests

joemcgill commented:
Found and made a couple small corrections and this is ready for another review.

joemcgill committed:
Merge branch 'update/2582-claim-ads-account' into update/2603-create-ads-account

joemcgill committed:
Limit Ads claim updates to 3 retries

ankitguptaindia commented:
**QA/Test Report-** ✅ 
------------


**Test Results -**  The Google Ads new account creation flow, account claiming, new account connection, and onboarding flow when using a new account are working as described in the requirements, and all tests have passed.

**Quick Demo:**  

https://github.com/user-attachments/assets/a7f35833-b3a3-4254-8dfa-07a6a55dc27e

eason9487 commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -1,122 +1,54 @@
  *
+ * @param {Object} props Component props.
+ * @param {Function} props.onRequestCreate A callback to fire when creating a new account.
+ * @param {boolean} props.upsertingAction Whether the user is in the process of creating/connecting a new Ads account.
```
The type is incorrect.

https://github.com/woocommerce/google-listings-and-ads/blob/d074504253e2c10c91bf944f062bcd51efb80727/js/src/hooks/useUpsertAdsAccount.js#L32

https://github.com/woocommerce/google-listings-and-ads/blob/d074504253e2c10c91bf944f062bcd51efb80727/js/src/hooks/useUpsertAdsAccount.js#L45

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
Could this PR not include code changes that are not attributed to it? The d074504253e2c10c91bf944f062bcd51efb80727 commit belongs more to #2644 (be3e8ec2f623f608c1554771789797aa7fdefacf).

joemcgill replied:
I actually made the commit originally in this PR, since testing all of the claim functionality properly was better when I could create a new account from the same UI, but I agree that functionally it fit more in the PR for the Claim functionality, so I cherry-picked this commit there. Given that the other will be merged first, it should be fine to leave this as is and resolve any merge conflicts after #2644 is merged.

eason9487 commented on js/src/components/google-combo-account-card/connect-ads/connect-ads-footer.js:
```diff
@@ -14,17 +14,21 @@ import DisconnectAccount from '.~/components/google-ads-account-card/disconnect-
  *
  * @param {Object} props Props.
  * @param {boolean} props.isConnected Whether the account is connected.
+ * @param {Function} props.onCreateNewClick Callback to create a new account.
```
```suggestion
 * @param {Function} props.onCreateNewClick Callback when clicking on the button to create a new account.
```

eason9487 commented on js/src/components/google-combo-account-card/connect-ads/connect-existing-account.js:
```diff
@@ -0,0 +1,125 @@
+ * Renders an account card to connect to an existing Google Ads account.
+ *
+ * @param {Object} props Component props.
+ * @param {Function} props.onCreateClick A callback to fire when creating a new account.
```
```suggestion
 * @param {Function} props.onCreateClick Callback when clicking on the button to create a new account
```

eason9487 reviewed (approved):
There are three comments about JSDoc that need adjustments, and I believe they don't need a new round of code reviews. I will be approving it in advance.

💡 Please make sure that this PR will be merged after #2644 is merged to avoid code changes being overlapped with #2644.

joemcgill committed:
Inline docs updates

Co-authored-by: Eason <eason.su.tw@gmail.com>

joemcgill committed:
Docblock fix

joemcgill reviewed (changes requested):
Just marking this as needing changes until after #2644 is finalized and merged.

eason9487 commented on js/src/components/google-combo-account-card/connect-ads/connect-ads.js:
```diff
@@ -1,122 +1,54 @@
  *
+ * @param {Object} props Component props.
+ * @param {Function} props.onRequestCreate A callback to fire when creating a new account.
+ * @param {string} props.upsertingAction The action the user is performing. Either 'create' or 'update'.
```
Continue on https://github.com/woocommerce/google-listings-and-ads/pull/2651#discussion_r1837998702.

The `null` type is missing.

```suggestion
 * @param {string|null} props.upsertingAction The action the user is performing. Possible values are 'create', 'update', or null.
```

asvinb replied:
Thanks @eason9487 . JSDoc updated.

asvinb committed:
Update JSDoc.

joemcgill committed:
Merge branch 'feature/2509-consolidate-google-account-cards' into update/2603-create-ads-account

joemcgill committed:
E2E: Remove flaky mocks from auto account creation test

joemcgill reviewed (approved):
I've updated this to resolve conflicts after #2644 was merged. 

I then checked the relevant E2E tests in `tests/e2e/specs/setup-mc/step-1-accounts.test.js` and noticed that the existing tests are broken in the target branch due to the way mocks were being handled. I've fixed this test in f2aa1c98 to confirm those tests are once again passing once this is merged.

Additional improvements to the changed E2E test can be handled in a follow-up PR.
